src/v/bytes/bytes.h

```diff
@@ -16,22 +16,182 @@
-  >;
-
-using bytes_view = std::basic_string_view<uint8_t>;
+
+class bytes {
+    using container_type
+      = ss::basic_sstring<char, uint32_t, bytes_inline_size, false>;
+
+public:
+    using value_type = uint8_t;
+    using size_type = uint32_t;
+    using reference = value_type&;
+    using const_reference = const value_type&;
+    using pointer = value_type*;
+    using const_pointer = const value_type*;
+    using iterator = value_type*;
+    using const_iterator = const value_type*;
+
+    static bytes from_string(std::string_view s) { return bytes(s); }
+
+    bytes() = default;
+    bytes(const bytes&) = default;
+    bytes& operator=(const bytes&) = default;
+    bytes(bytes&&) noexcept = default;
+    bytes& operator=(bytes&&) noexcept = default;
+    ~bytes() = default;
+
+    struct initialized_later {};
+    bytes(initialized_later, size_t size)
+      : data_(container_type::initialized_later{}, size) {}
+
+    struct initialized_zero {};
+    bytes(initialized_zero, size_t size)
+      : data_(size, 0) {}
+
+    bytes(const value_type* data, size_t size)
+      : data_(cast_down(data), size) {}
+
+    bytes(std::initializer_list<uint8_t> x)
+      : data_(cast_down(x.begin()), x.size()) {}
+
+    bytes(const value_type* begin, const value_type* end)
+      : data_(begin, end) {}
+
+    template<typename InputIterator>
+    bytes(InputIterator begin, InputIterator end)
+      : data_(begin, end) {}
+
+    explicit bytes(bytes_view);
+
+    reference operator[](size_type pos) noexcept {
+        return *cast_up(&data_[pos]);
+    }
+    const_reference operator[](size_type pos) const noexcept {
+        return *cast_up(&data_[pos]);
+    }
+
+    pointer data() noexcept { return cast_up(data_.data()); }
+    const_pointer data() const noexcept { return cast_up(data_.data()); }
+
+    iterator begin() noexcept { return cast_up(data_.begin()); }
+    const_iterator begin() const noexcept { return cast_up(data_.begin()); }
+    const_iterator cbegin() const noexcept { return cast_up(data_.cbegin()); }
+
+    iterator end() noexcept { return cast_up(data_.end()); }
+    const_iterator end() const noexcept { return cast_up(data_.end()); }
+    const_iterator cend() const noexcept { return cast_up(data_.cend()); }
+
+    size_type size() const noexcept { return data_.size(); }
+    bool empty() const noexcept { return data_.empty(); }
+
+    void resize(size_type size) { data_.resize(size); }
+
+    void append(const_pointer s, size_t n) { data_.append(cast_down(s), n); }
+
+    friend bool operator==(const bytes&, const bytes&) = default;
+
+    friend bool operator<(const bytes& a, const bytes& b) {
+        return a.data_ < b.data_;
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const bytes& b);
+
+private:
+    explicit bytes(std::string_view s)
+      : data_(s) {}
+
+    static const char* cast_down(const uint8_t* p) {
+        // NOLINTNEXTLINE
+        return reinterpret_cast<const char*>(p);
```

Comment:
I am pretty sure this general approach is UB: i.e., accessing an array of `char` (which is what `ss::string` contains) through a `uint8_t *` is UB, since they are different types. Except in very limited cases you can't convert a pointer of one type to a pointer to an unrelated type and then access though it (and even fewer cases when the pointers are to arrays).

That said, it's probably the type of UB that maybe works in practice?

Reply:
Interesting, thanks for the UB call out. TBH I thought that as long as the string is treated as a bag of bytes it was ok. I'll do some investigation. It would be nice if everything is above board!

Reply:
> So often in serialization/deserialization, though, we have a bag of bytes (e.g. char*) with a particular encoding which we can use reinterpret_cast to access. Is it that reinterpret_cast is always UB?

Not exactly, it's accessing an object of type `T` thought a pointer to type `U` that is UB, simply doing the cast itself isn't UB. Of course, to get a `U *` which actually points to `T` may require something like a reinterpret case (though there are other ways too: C-style cast, 2x static_cast though `void *`, memcpy etc.

This is the so-called "strict aliasing rules".

> we have a bag of bytes

There is an exception for `char`, but it kind of only works one way. For any type, you can inspect and write it's bytes using `char *`. So the following (access `int` through a `char` pointer) is valid:

```
int some_int = 5;
char * as_bytes = reinterpret_cast<char *>(&some_int);
printf("byte 2 is %d", as_bytes[2]);
```

but the reverse (access char through int pointer) is not:

```
char some_chars[] = {1, 2, 3, 4};
int * as_int = reinterpret_cast<int *>(&some_chars);
printf("four bytes as int: %d", *as_int);
```

In this case we are least sometimes doing the "reverse" (disallowed) case because `ss::string` puts chars into the array then we access them as `uint8_t`.

However, it seems quite unlikely this UB will bite use in practice:

1) The aliasing exception above for `char` applies to all "character types", which includes `unsigned char` too. uint8_t can _in principle_ be a different type from `char`/`unsigned char` (i.e., _not_ a character type) but in practice it is `unsigned char`, so in fact the aliasing exception applies. It's still "weird" that `ss::string` is treating it as `char` and the wrapper as `unsigned char`, but this in the realm of unspecified behavior (depends on the representation of char and uchar but we know they are the same), not undefined behavior.

2) Even though the aliasing except is "one way" per above, compilers have a pretty hard time applying that aspect and so IME as long as one of the types is a char type, it's not going to do tricky aliasing related optimizations even when you do the "reverse" (disallowed) case. I couldn't come up with any example where it does, anyway.

Godbot:

https://godbolt.org/z/jYjPq4rPo

`add_int` shows that strict aliasing is used by the optimizer: `i0 + ints[0]` is effectively collapsed to `2 * ints[0]` (i.e., `ints` is read only once) even though there is an intervening write of the `shorts` array: the compiler knows `shorts` can't alias `ints` because the they different types. `add_char` shows the opposite: the optimization is not applied because at least one side (in this case both) are character-types, so the aliasing exception applies.